### PR TITLE
Vtol rework

### PR DESCRIPTION
**Pull request by tumbili**
- move tailsitter and tiltrotor to specific classes
- vtol attitude control module should only contain generic function calls but no airframe specific code

### CODE REVIEW

**Comment by tumbili** [comment]
@LorenzMeier This is the best way I have found to integrate both changes from the FireFly and the Caipirinha into master. I don't think that parameters inside vtol_attitude_control module would have been nice, because the control of the two airframe types differs in many aspects.
My goal was to have a very simple structure in vtol_att_control_main which is very similar to the one before (have fw and mc in parallel). Then you have generic function calls to the airframe class where tailsitter/tiltorotor can do whatever they want to.

File: src/modules/vtol_att_control/airframe.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	virtual void update_vtol_state() = 0;
+	virtual void update_mc_state() = 0;
+	virtual void process_mc_data() = 0;
+	virtual void update_fw_state() = 0;
+	virtual void process_fw_data() = 0;
+	virtual void update_transition_state() = 0;
+	virtual void update_external_state() = 0;
+};
+
+#endif
```
Comment (tumbili) [inline src/modules/vtol_att_control/airframe.h]: newline

File: src/modules/vtol_att_control/tailsitter.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+{
+	_actuators_out_0.control[0] = _actuators_mc_in.control[0];
+	_actuators_out_0.control[1] = _actuators_mc_in.control[1];
+	_actuators_out_0.control[2] = _actuators_mc_in.control[2];
+	_actuators_out_0.control[3] = _actuators_mc_in.control[3];
+	//set neutral position for elevons
+	_actuators_out_1.control[0] = _actuators_mc_in.control[2];	//roll elevon
+	_actuators_out_1.control[1] = _actuators_mc_in.control[1];;	//pitch elevon
+	_actuators_out_1.control[4] = _tilt_control;	// for tilt-rotor control
+}
```
Comment (tumbili) [inline src/modules/vtol_att_control/tailsitter.cpp]: newline

File: src/modules/vtol_att_control/tailsitter.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	void update_external_state();
+
+	private:
+		void fill_mc_att_control_output();
+		void fill_fw_att_control_output();
+		void calc_tot_airspeed();
+		void scale_mc_output();
+	
+
+ };
```
Comment (tumbili) [inline src/modules/vtol_att_control/tailsitter.h]: newline

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		pwm_values.values[i] = 2000;
+		pwm_values.channel_count = _params.vtol_motor_count;
+	}
+
+	ret = ioctl(fd, PWM_SERVO_SET_MAX_PWM, (long unsigned int)&pwm_values);
+
+	if (ret != OK) {errx(ret, "failed setting max values");}
+
+	close(fd);
+}
```
Comment (tumbili) [inline src/modules/vtol_att_control/tiltrotor.cpp]: newline

File: src/modules/vtol_att_control/tiltrotor.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	void update_external_state();
+
+	private:
+
+	void fill_mc_att_control_output();
+	void fill_fw_att_control_output();
+	void set_max_mc();
+	void set_max_fw(unsigned pwm_value);
+
+ };
```
Comment (tumbili) [inline src/modules/vtol_att_control/tiltrotor.h]: newline

File: src/modules/vtol_att_control/vtol_att_control_main.h [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	void 		vehicle_battery_poll();			// Check for battery updates
+	void 		parameters_update_poll();		//Check if parameters have changed
+	int 		parameters_update();			//Update local paraemter cache
+	void 		fill_mc_att_rates_sp();
+	void 		fill_fw_att_rates_sp();
+	void 		set_idle_fw();
+	void 		set_idle_mc();
+};
+
+#endif
```
Comment (tumbili) [inline src/modules/vtol_att_control/vtol_att_control_main.h]: newline

File: src/modules/vtol_att_control/vtol_att_control_params.c [hunk 7]
```diff
@@ -142,3 +142,11 @@ PARAM_DEFINE_FLOAT(VT_PROP_EFF,0.0f);
...
 PARAM_DEFINE_FLOAT(VT_ARSP_LP_GAIN,0.3f);
 
+/**
+ * Aiframe type
+ *
+ * @min 0
+ * @max 1
+ * @group VTOL Attitude Control
+ */
+PARAM_DEFINE_INT32(VT_AIRFRAME, 0);
```
Comment (tumbili) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: newline

File: ROMFS/px4fmu_common/init.d/13002_firefly6 [hunk 8]
```diff
@@ -41,3 +41,4 @@ set MAV_TYPE 21
 
 param set VT_MOT_COUNT 6
 param set VT_IDLE_PWM_MC 1080
+param set VT_AIRFRAME 1
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/13002_firefly6]: Newline missing.

File: ROMFS/px4fmu_common/init.d/13003_tailsitter_x [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+set MIXER tailsitter_x
+
+set PWM_OUT 1234
+set PWM_MAX 2000
+set PWM_RATE 400
+set MAV_TYPE 20
+param set VT_MOT_COUNT 4
+param set VT_IDLE_PWM_MC  1080
+param set VT_AIRFRAME 0
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/13003_tailsitter_x]: Newline missing.

File: ROMFS/px4fmu_common/mixers/tailsitter_x.main.mix [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#mixer for the elevons
+M: 2
+O:      10000  10000      0 -10000  10000
+S: 1 0  10000  10000      0 -10000  10000
+S: 1 1  10000  10000      0 -10000  10000
+
+M: 2
+O:       10000   10000    0 -10000  10000
+S: 1 0   10000   10000    0 -10000  10000
+S: 1 1  -10000  -10000    0 -10000  10000
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/mixers/tailsitter_x.main.mix]: Newline missing.

File: src/modules/vtol_att_control/airframe.cpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+#include "airframe.h"
+
+Airframe::Airframe() {
+
+}
+
+Airframe::~Airframe() {
+	
+}
```
Comment (LorenzMeier) [inline src/modules/vtol_att_control/airframe.cpp]: Newline missing..

File: src/modules/vtol_att_control/tailsitter.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+ /**
+ * @file tiltrotor.h
+ *
+ * @author Roman Bapst 		<bapstroman@gmail.com>
+ *
+ */
+#include "vtol_att_control_main.h"
+#include "airframe.h"
+
+ class Tailsitter : public VtolAttitudeControl, public Airframe
```
Comment (LorenzMeier) [inline src/modules/vtol_att_control/tailsitter.h]: Multi-Inheritance is an evil pattern and generally better to be avoided - do you really need this here? Is the Tailsitter class an Airframe config or an Attitude controller? Why can't VTOLAttitudeControl just inherit from Airframe?

File: src/modules/vtol_att_control/vtol_att_control_main.cpp [hunk 13]
```diff
@@ -276,9 +119,22 @@ VtolAttitudeControl::VtolAttitudeControl() :
 	_params_handles.power_max = param_find("VT_POWER_MAX");
 	_params_handles.prop_eff = param_find("VT_PROP_EFF");
 	_params_handles.arsp_lp_gain = param_find("VT_ARSP_LP_GAIN");
+	_params_handles.airframe_type = param_find("VT_AIRFRAME");
```
Comment (LorenzMeier) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp]: Is the AIRFRAME here not rather encoding the type?

File: src/modules/vtol_att_control/vtol_att_control_params.c (hunk 7 above)
Comment (LorenzMeier) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: Looks like this is a type, not a specific airframe.

**Comment by LorenzMeier** [comment]
Looks good in general. I would recommend you to fix small things like newlines just immediately yourself rather than spending the time to comment them - its tedious, but much faster.

The only architectural change that is needed is to drop the multiple inheritance. There are many potential problems and confusion associated to it, and I don't see why it would be needed here. Either have the attitude controller as a member of the VTOL airframe class, or just inherit from the Airframe class.
http://stackoverflow.com/questions/406081/why-should-i-avoid-multiple-inheritance-in-c

File: src/modules/vtol_att_control/vtol_att_control_params.c (hunk 7 above)
Comment (tumbili) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: Yes, you are right.

**Comment by tumbili** [comment]
@LorenzMeier Yeah I understand. I can make the Airframe class inherit form VTOL_att_control and then let tailsitter and tiltrotor inherit from the airframe class.

**Comment by tumbili** [comment]
@LorenzMeier 
- removed the double inheritance: I now have a vtol_type class which inherits from vtol_att_control class. Then the tailsitter and tilrotor class both inherit from the vtol_type class.
- renamed airframe class to vtol_type class and also changed the VT_AIRFRAME param to VT_TYPE

**Comment by tumbili** [comment]
Closing in favour of #2418

**Pull request closed without merging**